app/config/Switches.scala

```diff
@@ -1,35 +1,29 @@
 package config
 
 import java.util.concurrent.{Executors, TimeUnit}
-
 import _root_.utils.Notifier
-import akka.agent.Agent
 import com.amazonaws.services.s3.AmazonS3
 import com.amazonaws.services.s3.model.{GetObjectRequest, ObjectMetadata, PutObjectRequest}
 import com.amazonaws.util.StringInputStream
 import utils.Loggable
-import org.quartz._
 import play.api.libs.json.{Format, JsString, JsValue, Json}
 
+import java.util.concurrent.atomic.AtomicReference
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.io.Source
 
 class Switches(config: LoginConfig, s3Client: AmazonS3) extends Loggable {
```

Comment:
Unfamiliar with this concept of switches in the login tool. Would it be possible to go over what they are used?

Reply:
AFAICS it's a way of –

- modelling the emergency switch, which is either on/off, in a threadsafe way (the fact that our map supports other switches is at the moment redundant), and providing an API to flip it
- persisting that state to S3, and providing a mechanism to ensuring that other boxes are up to date with what's persisted given the window defined in the scheduler for `refresh()`

Reply:
That explains the complexity. Thanks for clearing that up.